fl4health/strategies/feddg_ga_strategy.py

```diff
@@ -0,0 +1,382 @@
+            assert isinstance(config["n_server_rounds"], int), "n_server_rounds is not an integer"
+            self.num_rounds = config["n_server_rounds"]
+
+        return client_fit_ins
+
+    def aggregate_fit(
+        self,
+        server_round: int,
+        results: List[Tuple[ClientProxy, FitRes]],
+        failures: List[Union[Tuple[ClientProxy, FitRes], BaseException]],
+    ) -> Tuple[Optional[Parameters], Dict[str, Scalar]]:
+        """
+        Aggregate fit results by weighing them against the adjustment weights and then summing them.
+
+        Collects the fit metrics that will be used to change the adjustment weights for the next round.
+
+        Args:
+            server_round: (int) the current server round.
+            results: (List[Tuple[ClientProxy, FitRes]]) The clients' fit results.
+            failures: (List[Union[Tuple[ClientProxy, FitRes], BaseException]]) the clients' fit failures.
+
+        Returns:
+            (Tuple[Optional[Parameters], Dict[str, Scalar]]) A tuple containing the aggregated parameters
+                and the aggregated fit metrics.
+        """
+
+        _, metrics_aggregated = super().aggregate_fit(server_round, results, failures)
+
+        self.train_metrics = {}
+        for result in results:
+            self.train_metrics[result[0].cid] = result[1].metrics
+
+        parameters_aggregated = ndarrays_to_parameters(self.weight_and_aggregate_results(results))
+
+        return parameters_aggregated, metrics_aggregated
+
+    def aggregate_evaluate(
+        self,
+        server_round: int,
+        results: List[Tuple[ClientProxy, EvaluateRes]],
+        failures: List[Union[Tuple[ClientProxy, EvaluateRes], BaseException]],
+    ) -> Tuple[Optional[float], Dict[str, Scalar]]:
+        """
+        Aggregate evaluation losses using weighted average.
+
+        Collects the evaluation metrics and update the adjustment weights, which will be used
+        when aggregating the results for the next round.
+
+        Args:
+            server_round: (int) the current server round.
+            results: (List[Tuple[ClientProxy, FitRes]]) The clients' evaluate results.
+            failures: (List[Union[Tuple[ClientProxy, FitRes], BaseException]]) the clients' evaluate failures.
+
+        Returns:
+            (Tuple[Optional[float], Dict[str, Scalar]]) A tuple containing the aggregated evaluation loss
+                and the aggregated evaluation metrics.
+        """
+        loss_aggregated, metrics_aggregated = super().aggregate_evaluate(server_round, results, failures)
+
+        self.evaluation_metrics = {}
+        for result in results:
+            cid = result[0].cid
+            self.evaluation_metrics[cid] = result[1].metrics
+            # adding the loss to the metrics
+            val_loss_key = FairnessMetricType.LOSS.value
+            self.evaluation_metrics[cid][val_loss_key] = loss_aggregated  # type: ignore
```

Comment:
Out of curiosity, why do we need the type ignore here?

Reply:
Because `loss_aggregated` can apparently be None (type is `Optional[float]`) and mypy does't like that being assigned to `Scalar`.

Reply:
Ah right, I see. If the value is, in fact, None, does other stuff break? That is, should we just type narrow by asserting that `loss_aggregated is not None` here?

Reply:
I was assuming if loss is None we will probably have bigger problems but I guess it doesn't hurt to add an assertion here.